### PR TITLE
[Trainer] FP4, 8, 16 training by native dtype support and quant linear autograd function

### PR DESCRIPTION
**Pull request by KohakuBlueleaf**
In this pull request I proposed native dtype training and quant linear bwd support for faster training:

1. Now the training node support native dtype training, it will NOT force bf16 or fp32 compute but utilize ComfyUI native dtype handling system. To achieve this, the training node will automatically detect if grad scaler is needed (for fp16 compute dtype), and automatically apply it
2. Now the training node support "quantized model training" (only fp8/nvfp4 linear model for now), with bypass mode enabled you can train Flux Klein 4B nvfp4 with 768x768 resolution on local RTX4090 easily. (bs1, around 1.3 it/sec on 4090 with bf16 manual cast) To achieve this I wrapped the original quantized forward part with an autograd function, in backward computation I dequant the weight back to compute dtype and compute backward gradient.

Some limitation:
1. For native dtype training, FP16 + FP16 accumulation may lead to unstable training (loss=Nan), disable fp16 accumulation can resolve this.
2. For native dtype + quantized model. Currently only "fp8, nvfp4 on linear only" supported. And bypass mode is required.
3. Quantized backward (like fp8e4m3 fwd + fp8e5m2 bwd or fp4 fwd + fp8e5m2 bwd) is currently not supported, a future PR may add this kind of feature if we have good work around on handling different quant for single model.

Example (Flux Klein 4B nvfp4, trained on RTX4090 with 0.6M pixel with style dataset)
image

### CODE REVIEW

File: comfy_extras/nodes_train.py [hunk 1]
```diff
@@ -1120,22 +1136,32 @@ def execute(
...
+                # Warn about fp16 accumulation instability during training
+                if PerformanceFeature.Fp16Accumulation in args.fast:
+                    logging.warning(
+                        "WARNING: FP16 model detected with fp16_accumulation enabled. "
+                        "This combination can be numerically unstable during training and may cause NaN values. "
+                        "Suggested fixes: 1) Set training_dtype to 'bf16', or 2) Disable fp16_accumulation (remove from --fast flags)."
+                    )
+            else:
+                # For fp8, bf16, or other dtypes, use bf16 autocast
+                dtype = torch.bfloat16
```
[the remark this answers is not in the document]
Comment (KohakuBlueleaf) [reply comfy_extras/nodes_train.py:1158]: Expected behavior.

File: comfy/ops.py [hunk 2]
```diff
@@ -689,6 +689,71 @@ def forward(self, *args, **kwargs):
...
+
+        w = weight.detach() if weight.requires_grad else weight
+        b = bias.detach() if bias is not None and bias.requires_grad else bias
+
+        output = torch.nn.functional.linear(q_input, w, b)
+
+        # Restore original input shape
+        if len(input_shape) > 2:
+            output = output.unflatten(0, input_shape[:-1])
+
```
[the remark this answers is not in the document]
Comment (KohakuBlueleaf) [reply comfy/ops.py:716]: 1D tensor will not exist for our usecase
